### PR TITLE
Final borrow

### PR DESCRIPTION
**Pull request by arnaudgolfouse**
Experiment to repair proofs when using a third field in mutable borrows.

In short, in the following code:
```rust
let mut x = 0i32;
let mut y = 0i32;
let ref_x_1 = &mut x;
let ref_x_2 = &mut *ref_x_1;
let ref_y_1 = &mut y;
let ref_y_2 = &mut *ref_y_1;

// ref_x_1 is never used here

let z = *ref_y_1;
```

Then:
- The third field of `ref_x_1` and `ref_x_2` should be the same, because `ref_x_1` is "dead" after the reborrow (we say that this reborrow is **final**)
- This is not the case for `ref_y_1`, so the third field of `ref_y_1` and `ref_y_2` is not the same.

### CODE REVIEW

**Comment by arnaudgolfouse** [comment]
Some tests had their source (`.rs`) changed:
- `iterators/02_iter_mut`:
  - in `completed`: `self.resolve()` becomes `(*self).inner@ == (^self).inner@`
  - in `produces`, rather than comparing the two sequences directly, we need to compare each borrow by their current/final value.
- `iterators/06_map_precond`: remove `let _ = self;`
- `take_first_mut`: replace `^self == *self` with `(*^self_)@.len() == 0 && (^*self_)@.len() == 0`
   
   We cannot prove `^self == *self`, because we used `std::mem::take`, and the new borrow may not have the same id.

**Comment by jhjourdan** [comment]
Good !

First, CI is failing, which is indicating some of the sessions need to be updated. And indeed, it says that the iter_mut session is obsolete. Could you please update it?

> in completed: self.resolve() becomes (*self).inner@ == (^self).inner@

This is suspicious. Why is this necessary? It should be equivalent, right?

> `iterators/06_map_precond`: remove `let _ = self;`

I confirm that this is OK. This dates from the time where we had an assertion that we needed to prove the invariant. This assertion disappeared, so we can remove this hack.

That being being said, this is a real problem: we may need a way to reliably delay resolution without breaking final reborrows...

**Comment by xldenis** [comment]
> That being being said, this is a real problem: we may need a way to reliably delay resolution without breaking final reborrows...

Let's not complicate an already complicated change with hypothetical edge-cases. The day when this becomes necessary we can revisit it.

**Comment by xldenis** [comment]
> This is suspicious. Why is this necessary? It should be equivalent, right?

That's really weird, if this is breaking it also would indicate deeper issues to me. since `^self = *self` should imply that.

**Comment by jhjourdan** [comment]
> Let's not complicate an already complicated change with hypothetical edge-cases. The day when this becomes necessary we can revisit it.

Yes, of course! I was mentioning this simply because we may need to do something at some point.

File: creusot/src/analysis/not_final_places.rs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+where
+    T: GenKill<PlaceId>,
+{
+    fn visit_place(&mut self, place: &Place<'tcx>, context: mir::visit::PlaceContext, _: Location) {
+        use mir::visit::{MutatingUseContext, PlaceContext};
+        match (self.context, context) {
+            // TODO: should we really accept _all_ nonmutating uses ?
+            (Context::Read, PlaceContext::NonMutatingUse(_))
+            // Note that a borrow is considered a read
+            | (Context::Read, PlaceContext::MutatingUse(MutatingUseContext::Borrow)) => {
```
[the remark this answers is not in the document]
Comment (jhjourdan) [reply creusot/src/analysis/not_final_places.rs:257]: Same thing for `Call` and `SetDiscriminant` bellow a deref.

**Review by jhjourdan — DISMISSED** [review]
All in all I think this is very good news and should be merged quickly. Thanks.

There are still quite a few changes needed. Namely:

- First, I think there are a few bugs which I detailed in the review.
- Cleanup sessions. A few sessions contains "timeout" entries for some provers. They should not be included for successful goals (Why3 shows a warning for them, I think)
- The iter_mut test has an obsolete session. The new shapes file should be comitted.
- Some session files contains duplicate provers used for the same goals. This is not really a problem here, but it slows down the CI, so you should be cautious about them, and identify what part of your workflow with Why3IDE makes these duplicated prover calls.
- Not all changes to iter_mut and take_first_mut have been propagated to `creusot_contracts`.
- @xldenis have merged #852, so you should resolve the conflicts, and add reborrowing of indexed projections

I'm sorry to ask you these changes, but you're almost there!

**Comment by jhjourdan** [comment]
> That's really weird, if this is breaking it also would indicate deeper issues to me. since `^self = *self` should imply that.

Actually it does not imply this, because @arnaudgolfouse's version does not say anything about the inner tag. And I think @arnaudgolfouse is right: his change is necessary because of the change of the specification of `take_first_mut`.

That being said, `(^self).inner@  == Seq::EMPTY` would have been clearer (same remark for `take_first_mut`, BTW).

File: creusot/src/analysis/not_final_places.rs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+        let mut visitor = VisitAllPlaces(HashMap::new());
+        visitor.visit_body(body);
+        let places = visitor.0;
+        let mut contains_dereference = BitSet::new_empty(places.len());
+        let mut subplaces = vec![Vec::new(); places.len()];
+        let mut conflicting_places = vec![Vec::new(); places.len()];
+        for (place, id) in &places {
+            if place.projection.get(1..).unwrap_or_default().contains(&ProjectionElem::Deref) {
```
[the remark this answers is not in the document]
Comment (jhjourdan) [reply creusot/src/analysis/not_final_places.rs:113]: BTW, isn't this place.is_indirect?

File: creusot/src/backend/program.rs [hunk 3]
```diff
@@ -622,6 +622,42 @@ impl<'tcx> Place<'tcx> {
...
+        vec![original_borrow], // TODO: should be "nested" (how ?)
+    );
+    // NOTE: no Deref should be present.
+    for proj in projection {
+        match proj {
+            ProjectionElem::Deref => {
+                unreachable!(
+                    "the final borrow analysis should have garanteed that only one `Deref` appears"
+                )
+            }
```
Comment (jhjourdan) [inline creusot/src/backend/program.rs:640]: This is still possible if this is the Deref of a Box

File: creusot/src/analysis/not_final_places.rs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    /// Run the analysis right **after** `location`, and determines if `place` is a final reborrow.
+    pub fn is_final_at(
+        cursor: &mut ResultsCursor<'_, 'tcx, NotFinalPlaces<'tcx>>,
+        place: &Place<'tcx>,
+        location: Location,
+    ) -> bool {
+        let deref_position =
+            match place.projection.iter().position(|p| matches!(p, ProjectionElem::Deref)) {
```
Comment (jhjourdan) [inline creusot/src/analysis/not_final_places.rs:137]: You should not consider Deref if the dereferenced type is `Box<T>`

File: creusot/src/analysis/not_final_places.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhjourdan) [reply creusot/src/analysis/not_final_places.rs:113]: You should not consider Deref if the dereferenced type is `Box<T>`

File: creusot-contracts/src/std/slice.rs [hunk 5]
```diff
@@ -270,11 +270,11 @@ extern_spec! {
...
             Some(r) => {
-                *r == (**self)[0] && ^r == (^*self)[0] &&
-                (**self)@.len() > 0 && (^*self)@.len() > 0 &&
-                (*^self)@ == (**self)@.tail() && (^^self)@ == (^*self)@.tail()
+                *r == (**self_)[0] && ^r == (^*self_)[0] &&
+                (**self_)@.len() > 0 && (^*self_)@.len() > 0 &&
+                (*^self_)@ == (**self_)@.tail() && (^^self_)@ == (^*self_)@.tail()
             }
-            None => ^self == * self && (**self)@.len() == 0
+            None => (*^self_)@.len() == 0 && (^*self_)@.len() == 0 && (**self_)@.len() == 0 && (^^self_)@.len() == 0
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot-contracts/src/std/slice.rs]: Ok, I thought `self_` was necessary, but it indeed is not

File: creusot-contracts/src/std/slice.rs [hunk 6]
```diff
@@ -376,6 +376,7 @@ impl<'a, T> Iterator for Iter<'a, T> {
     #[open]
     fn produces(self, visited: Seq<Self::Item>, tl: Self) -> bool {
         pearlite! {
+            // TODO: rewrite without to_ref_seq
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot-contracts/src/std/slice.rs]: Removed this comment

File: creusot/src/backend/program.rs [hunk 7]
```diff
@@ -622,6 +622,42 @@ impl<'tcx> Place<'tcx> {
     }
 }
 
+pub(crate) fn borrow_generated_id<V: std::fmt::Debug, T: std::fmt::Debug>(
+    original_borrow: Exp,
+    projection: &[ProjectionElem<V, T>],
+) -> Exp {
+    let mut borrow_id = Exp::Call(
+        Box::new(Exp::pure_qvar(QName::from_string("Borrow.get_id").unwrap())),
+        vec![original_borrow], // TODO: should be "nested" (how ?)
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/backend/program.rs]: Nothing, that was just a reminder for something I already did, so I removed this comment

File: creusot/src/backend/place.rs [hunk 8]
```diff
@@ -207,40 +221,38 @@ pub(crate) fn translate_rplace<'tcx>(
...
-                        Box::new(Exp::impure_qvar(names.accessor(*id, subst, 0, *ix))),
+                        Box::new(QVar(names.accessor(*id, subst, 0, ix), purity)),
                         vec![inner],
                     );
                 }
                 e => unreachable!("{:?}", e),
             },
             Downcast(_, _) => {}
             Index(ix) => {
-                // TODO: Use [_] syntax
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/backend/place.rs:223]: done

File: creusot/src/backend/term.rs [hunk 9]
```diff
@@ -335,10 +354,11 @@ impl<'tcx> Lower<'_, 'tcx> {
     }
 }
 
+use crate::{backend::place::translate_rplace_gen, translation::projection_vec::map_projections};
 use rustc_hir::def_id::DefId;
 use rustc_middle::ty::{subst::SubstsRef, TyCtxt};
 
-use super::{dependency::Dependency, Why3Generator};
+use super::{dependency::Dependency, program::borrow_generated_id, Why3Generator};
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/backend/term.rs]: 👍

File: creusot/tests/should_succeed/iterators/02_iter_mut/why3session.xml [hunk 10]
```diff
@@ -3,99 +3,112 @@
...
   <goal name="next&#39;vc.1" expl="type invariant" proved="true">
-  <proof prover="3"><result status="valid" time="0.123258" steps="18585"/></proof>
+  <proof prover="3"><result status="valid" time="0.123258" steps="18256"/></proof>
   </goal>
   <goal name="next&#39;vc.2" expl="postcondition" proved="true">
-  <proof prover="4"><result status="valid" time="0.430760" steps="5876"/></proof>
+  <proof prover="0"><result status="timeout" time="1.000000" steps="3742094"/></proof>
+  <proof prover="1"><result status="timeout" time="1.000000" steps="533704"/></proof>
+  <proof prover="3"><result status="timeout" time="1.000000" steps="158566"/></proof>
+  <proof prover="4"><result status="timeout" time="1.000000" steps="24732"/></proof>
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/tests/should_succeed/iterators/02_iter_mut/why3session.xml]: done

File: creusot/src/analysis/not_final_places.rs [hunk 11]
```diff
@@ -0,0 +1,279 @@
+use std::collections::HashMap;
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs:6]: Those are also present in the other `analysis/*` files: I assumed this was necessary for using specific rustc libraries, is this not the case ?

File: creusot/tests/should_succeed/bug/874/why3session.xml [hunk 12]
```diff
@@ -27,13 +30,13 @@
...
+  <proof prover="3"><result status="timeout" time="1.000000" steps="17714"/></proof>
+  <proof prover="4"><result status="timeout" time="1.000000" steps="223747"/></proof>
   <transf name="split_vc" proved="true" >
    <goal name="can_extend&#39;vc.6.0" expl="assertion" proved="true">
-   <transf name="use_th" proved="true" arg1="seq.FreeMonoid">
-    <goal name="can_extend&#39;vc.6.0.0" expl="assertion" proved="true">
-    <proof prover="3"><result status="valid" time="0.478130" steps="14221"/></proof>
-    </goal>
-   </transf>
+   <proof prover="0" timelimit="10"><result status="valid" time="9.474103" steps="20751146"/></proof>
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/tests/should_succeed/bug/874/why3session.xml]: I think I broke it somehow at some point, it indeed works very well with "use_th seq.FreeMonoid" -> Alt-Ergo

File: creusot/tests/should_succeed/bug/874/why3session.xml [hunk 13]
```diff
@@ -27,13 +30,13 @@
   <proof prover="3"><result status="valid" time="0.010587" steps="63"/></proof>
   </goal>
   <goal name="can_extend&#39;vc.6" expl="assertion" proved="true">
+  <proof prover="0" memlimit="1000"><result status="timeout" time="1.000000" steps="3591024"/></proof>
+  <proof prover="1"><result status="timeout" time="1.000000" steps="253865"/></proof>
+  <proof prover="3"><result status="timeout" time="1.000000" steps="17714"/></proof>
+  <proof prover="4"><result status="timeout" time="1.000000" steps="223747"/></proof>
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/tests/should_succeed/bug/874/why3session.xml]: done

File: creusot/src/backend/place.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/backend/place.rs:223]: done

File: creusot/src/analysis/not_final_places.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jhjourdan) [reply creusot/src/analysis/not_final_places.rs:6]: Some of these files are more or less direct copy-paste from Rustc code, so we need to keep the licence. But your code is really original.

[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs:6]: Ah, ok I misunderstood the meaning of these comments, I'll remove them.

File: creusot/src/backend/program.rs (hunk 3 above)
Comment (arnaudgolfouse) [reply creusot/src/backend/program.rs]: Indeed, I changed some things in `NotFinalPlaces::is_final_at` to handles boxes, let me know if this is satisfactory

**Comment by jhjourdan** [comment]
Could you please do a pass on each of my comments, check you have indeed handle them, and mark them as resolved? Thanks !

File: creusot/src/analysis/not_final_places.rs (hunk 4 above)
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs]: This should now be correctly handled, see `NotFinalPlaces::has_indirection`

File: creusot/src/analysis/not_final_places.rs [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        match (self.context, context) {
+            // TODO: should we really accept _all_ nonmutating uses ?
+            (Context::Read, PlaceContext::NonMutatingUse(_))
+            // Note that a borrow is considered a read
+            | (Context::Read, PlaceContext::MutatingUse(MutatingUseContext::Borrow)) => {
+                self.trans.gen(self.mapping.places[&place.into()])
+            }
+            (
+                Context::Write,
+                PlaceContext::MutatingUse(MutatingUseContext::Drop | MutatingUseContext::Store),
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs]: Yep, done

File: creusot/src/analysis/not_final_places.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs]: Yes, I should handle all cases correctly now (with tests!)

File: creusot/src/analysis/not_final_places.rs [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+        let mut conflicting_places = vec![Vec::new(); places.len()];
+        for (place, id) in &places {
+            if place.projection.get(1..).unwrap_or_default().contains(&ProjectionElem::Deref) {
+                contains_dereference.insert(*id);
+            }
+            for (other_place, other_id) in &places {
+                if id == other_id || place.local != other_place.local {
+                    continue;
+                }
+                if other_place.projection.get(..place.projection.len()) == Some(place.projection) {
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs:109]: I duplicated some of rustc's API here, this is the `places_conflict` function.

File: creusot/src/analysis/not_final_places.rs [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+            .visit_terminator(terminator, location)
+    }
+
+    fn call_return_effect(
+        &mut self,
+        trans: &mut impl GenKill<Self::Idx>,
+        _block: BasicBlock,
+        return_places: rustc_mir_dataflow::CallReturnPlaces<'_, 'tcx>,
+    ) {
+        return_places.for_each(|place| trans.gen(self.places[&(&place).into()]));
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs]: Alright, removed this

File: creusot/src/analysis/not_final_places.rs [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+/// ```
+#[derive(Debug)]
+pub struct NotFinalPlaces<'tcx> {
+    /// Mapping from a place to its ID
+    ///
+    /// This is necessary to use a `BitSet<PlaceId>`.
+    places: HashMap<PlaceAsSLice<'tcx>, PlaceId>,
+    /// Used to ban unnesting (`**ref_ref_x = ...`)
+    ///
+    /// In practice, this means the underlying place has no `Deref`.
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs]: Indeed, I reworked the comments

File: creusot/src/analysis/not_final_places.rs [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+    fmt::DebugWithContext, AnalysisDomain, Backward, GenKill, GenKillAnalysis, ResultsCursor,
+};
+
+pub type PlaceId = usize;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct PlaceAsSLice<'tcx> {
+    local: mir::Local,
+    projection: &'tcx [mir::PlaceElem<'tcx>],
+}
```
[the remark this answers is not in the document]
Comment (arnaudgolfouse) [reply creusot/src/analysis/not_final_places.rs]: Thanks!

File: creusot/src/backend/program.rs [hunk 19]
```diff
@@ -622,6 +622,42 @@ impl<'tcx> Place<'tcx> {
...
+                // since only one variant can be active at a time, there is no need to change the borrow index further
+            }
+            ProjectionElem::Index(_)
+            | ProjectionElem::ConstantIndex { .. }
+            | ProjectionElem::Subslice { .. }
+            | ProjectionElem::OpaqueCast(_) => {
+                todo!("{proj:?} not supported as final")
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jhjourdan) [reply creusot/src/backend/program.rs:676]: Ok, I was wrong.

**Pull request merged**